gorm/gorm.go

```diff
@@ -0,0 +1,222 @@
+package gorm
+
+import (
+	"database/sql"
+	"errors"
+	"time"
+
+	"gorm.io/driver/mysql"
+	"gorm.io/driver/postgres"
+	"gorm.io/driver/sqlite"
+	"gorm.io/driver/sqlserver"
+	"gorm.io/gorm"
+	"gorm.io/gorm/logger"
+)
+
+// Storage interface that is implemented by storage providers
+type Storage struct {
+	db         *gorm.DB
+	sqldb      *sql.DB
+	gcInterval time.Duration
+	tableName  string
+	done       chan struct{}
+}
+
+type table struct {
+	Key        string `gorm:"column:k;primaryKey;type:varchar(64);not null"`
+	Value      []byte `gorm:"column:v;not null"`
+	Expiration int64  `gorm:"column:e;index;not null"`
+}
+
+// New creates a new storage
+func New(config ...Config) *Storage {
+	var err error
+	var db *gorm.DB
+	var sqldb *sql.DB
+
+	// Set default config
+	cfg := configDefault(config...)
+
+	if cfg.Db != nil {
+		// Use passed db
+		db = cfg.Db
+
+		sqldb, err = db.DB()
+		if err != nil {
+			panic(err)
+		}
+	} else {
+		// Create db
+		switch cfg.Dialect {
```

Comment:
I'm wondering if we just need to support DB parameters, as adding other parameters isn't necessarily useful.

Reply:
Well, I think that every Storage package should only support a DB parameter initialized outside the package itself.
Many DBs have a lot config options that make it hard to fully support them, other times you just want to share your already initialized data connection for sessions too.
I'm not experienced enough to discuss precedent decisions, so I have just tried to copy the same code already in production but if you need it I can change it...
Maybe I can find some time to align all other Storage packages to support just an external connection pointer.